internal/storage/storageutil/client_helper.go

```diff
@@ -0,0 +1,110 @@
+	mountpkg "github.com/googlecloudplatform/gcsfuse/internal/mount"
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+)
+
+const CustomEndpoint = "https://localhost:9000"
+const DummyKeyFile = "test/test_creds.json"
+
+type StorageClientConfig struct {
+	ClientProtocol      mountpkg.ClientProtocol
+	MaxConnsPerHost     int
+	MaxIdleConnsPerHost int
+	HttpClientTimeout   time.Duration
+	MaxRetryDuration    time.Duration
+	RetryMultiplier     float64
+	UserAgent           string
+	Endpoint            *url.URL
+	KeyFile             string
+	TokenUrl            string
+	ReuseTokenFromUrl   bool
+}
+
+// GetDefaultStorageClientConfig is only for test, making the default endpoint
+// non-nil, so that we can create dummy tokenSource while unit test.
+func GetDefaultStorageClientConfig() (clientConfig StorageClientConfig) {
+	return StorageClientConfig{
+		ClientProtocol:      mountpkg.HTTP1,
+		MaxConnsPerHost:     10,
+		MaxIdleConnsPerHost: 100,
+		HttpClientTimeout:   800 * time.Millisecond,
+		MaxRetryDuration:    30 * time.Second,
+		RetryMultiplier:     2,
+		UserAgent:           "gcsfuse/unknown (Go version go1.20-pre3 cl/474093167 +a813be86df) (GCP:gcsfuse)",
+		Endpoint:            &url.URL{},
+		KeyFile:             DummyKeyFile,
+		TokenUrl:            "",
+		ReuseTokenFromUrl:   true,
+	}
+}
+
+func CreateHttpClientObj(storageClientConfig *StorageClientConfig) (httpClient *http.Client, err error) {
+	var transport *http.Transport
+	// Using http1 makes the client more performant.
+	if storageClientConfig.ClientProtocol == mountpkg.HTTP1 {
+		transport = &http.Transport{
+			MaxConnsPerHost:     storageClientConfig.MaxConnsPerHost,
+			MaxIdleConnsPerHost: storageClientConfig.MaxIdleConnsPerHost,
+			// This disables HTTP/2 in transport.
+			TLSNextProto: make(
+				map[string]func(string, *tls.Conn) http.RoundTripper,
+			),
+		}
+	} else {
+		// For http2, change in MaxConnsPerHost doesn't affect the performance.
+		transport = &http.Transport{
+			DisableKeepAlives: true,
+			MaxConnsPerHost:   storageClientConfig.MaxConnsPerHost,
+			ForceAttemptHTTP2: true,
+		}
+	}
+
+	tokenSrc, err := createTokenSource(storageClientConfig)
+	if err != nil {
+		err = fmt.Errorf("while fetching tokenSource: %w", err)
+		return
+	}
+
+	// Custom http client for Go Client.
+	httpClient = &http.Client{
+		Transport: &oauth2.Transport{
+			Base:   transport,
+			Source: tokenSrc,
+		},
+		Timeout: storageClientConfig.HttpClientTimeout,
+	}
+
+	// Setting UserAgent through RoundTripper middleware
+	httpClient.Transport = &userAgentRoundTripper{
+		wrapped:   httpClient.Transport,
+		UserAgent: storageClientConfig.UserAgent,
+	}
+
+	return httpClient, err
+}
+
+// IsProdEndpoint GCSFuse assumes an endpoint as prod endpoint
+// if user haven't specified via --endpoint flag.
+// Also, we don't encourage use --endpoint flag to pass actual gcs url
+// in that case, some configuration is not set. E.g. using MTLS.
```

Comment:
I didnt follow this comment. why can't users pass actual endpoint?

Reply:
Ref: https://screenshot.googleplex.com/89HEJFnx2HtvSuz

Reply:
I still didnt follow the comment - why does MTLS will get missed if prod endpoint is passed.

Reply:
I didn't look into much details but - somewhere we override the MTLS - https://github.com/googleapis/google-api-go-client/blob/f7f1c7b4c610f6a41c130c3bb85d6a609e99fb37/internal/cba.go#L98

by default MTLS endpoint - https://storage.mtls.googleapis.com/storage/v1/ but this gets override with endpoint.